rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,321 @@
+# RFC 8288 - 2021-07-20 - CSV Enrichment
+
+Vector needs to allow users to enrich the events flowing through the topology
+using a CSV file. This RFC proposes adding a VRL function that can perform
+this lookup.
+
+## Scope
+
+The RFC will cover a simple VRL function that will lookup a single row from a
+CSV file using a set of conditions. This allows the user to map the data into
+the event using the full power of VRL.
+
+For MVP purposes, lookup will be a simple sequential scan. The first version
+of this function will not handle any indexing of this data. Above a certain size
+indexing will provide substantial performance benefits, so it is something that
+we should definitely consider at a future stage. Therefore, it is out of scope
+for this RFC.
+
+Encryption will not be covered. This is very much a requirement to be considered
+at a future stage, but is out of scope for this RFC.
+
+Likewise, reloading the enrichment data is likely to be needed, but will not be
+covered here. For the first implementation of this RFC, reloading the data will
+require Vector to be restarted.
+
+## Motivation
+
+Users would like to enrich events flowing through Vector with extra data
+provided by a CSV file.
+
+## User Experience
+
+### Tables
+
+To represent the CSV file we have a new top level configuration option.
+
+```
+[tables.csv_file]
+  type = "csv"
```

Comment:
I'm in two minds about this. One one hand there would be a lot of shared fields between say a json and a csv file table, but there is one field - `delimiter` which would apply only to csv. I'll change it to a `file` type for now, but that may be liable to change back..

Reply:
This might be something where we just go with the specific version for now and then make a call about whether or not to generalize before the feature is actually released.

Reply:
Yeah, that's a good point. There will be options specific to the file type. We could probably still model that like:

```toml
type = "file"
encoding.type= "csv"
encoding.csv.delimiter = ","
encoding.csv.header_row = true
```

Where the CSV encoding options would only apply if `encoding.type = "csv"`.